File: src/main/java/gay/skitbet/jackiro/task/UpdateStatusTask.java
package gay.skitbet.jackiro.task;

import gay.skitbet.jackiro.Jackiro;
import gay.skitbet.jackiro.JackiroLauncher;
import gay.skitbet.jackiro.utils.JackiroShards;
import lombok.SneakyThrows;
import net.dv8tion.jda.api.OnlineStatus;
import net.dv8tion.jda.api.entities.Activity;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.lang.management.ManagementFactory;
import java.util.concurrent.TimeUnit;

public class UpdateStatusTask extends Thread {

    private static final Logger logger = LoggerFactory.getLogger(UpdateStatusTask.class);  // Logger for better logging
    private static final long UPDATE_INTERVAL = 25000L;  // 25 seconds
    private long lastUpdate = System.currentTimeMillis();
    private int currentIndex = 0;

    public UpdateStatusTask() {
        super("Update Status Task");
    }

    @SneakyThrows
    @Override
    public void run() {
        while (true) {
            try {
                updateStatus();
            } catch (Exception e) {
                logger.error("Error updating status: ", e);
            }
            Thread.sleep(1000);  // Sleep for 1 second before checking again
        }
    }

    private void updateStatus() {
        long uptime = ManagementFactory.getRuntimeMXBean().getUptime();
        String uptimeStr = formatUptime(uptime);

        String status = Jackiro.config.getCurrentlyPlaying().get(currentIndex);

        JackiroShards jackiroShards = Jackiro.getInstance().getJackiroShards();
        status = status.replace("{guilds}", String.valueOf(jackiroShards.getGuilds().size()));
        status = status.replace("{users}", String.valueOf(jackiroShards.getUsers().size()));
        status = status.replace("{uptime}", uptimeStr);

        long diff = System.currentTimeMillis() - lastUpdate;

        if (diff >= UPDATE_INTERVAL) {
            jackiroShards.setPresence(OnlineStatus.ONLINE, Activity.playing(status));
            currentIndex = (currentIndex + 1) % Jackiro.config.getCurrentlyPlaying().size();
            lastUpdate = System.currentTimeMillis();
        }
    }

    private String formatUptime(long uptime) {
        long days = TimeUnit.MILLISECONDS.toDays(uptime);
        uptime -= TimeUnit.DAYS.toMillis(days);
        long hours = TimeUnit.MILLISECONDS.toHours(uptime);
        uptime -= TimeUnit.HOURS.toMillis(hours);
        long minutes = TimeUnit.MILLISECONDS.toMinutes(uptime);
        uptime -= TimeUnit.MINUTES.toMillis(minutes);
        long seconds = TimeUnit.MILLISECONDS.toSeconds(uptime);

        return String.format("%dd %dh %dm %ds", days, hours, minutes, seconds);
    }
}

File: src/main/kotlin/gay/skitbet/jackiro/task/UpdateStatusTask.kt
package gay.skitbet.jackiro.task

import gay.skitbet.jackiro.Jackiro
import net.dv8tion.jda.api.OnlineStatus
import net.dv8tion.jda.api.entities.Activity
import org.slf4j.LoggerFactory
import java.lang.management.ManagementFactory
import java.util.concurrent.Executors
import java.util.concurrent.ScheduledExecutorService
import java.util.concurrent.TimeUnit

class UpdateStatusTask {

    private val logger = LoggerFactory.getLogger(UpdateStatusTask::class.java)
    private var currentIndex = 0
    private var lastUpdate = System.currentTimeMillis()

    companion object {
        private const val UPDATE_INTERVAL_MS = 25_000L
    }

    fun start(scheduler: ScheduledExecutorService) {
        scheduler.scheduleAtFixedRate(::updateStatusSafely, 0, 1, TimeUnit.SECONDS)
    }

    private fun updateStatusSafely() {
        try {
            updateStatus()
        } catch (e: Exception) {
            logger.error("Error updating status", e)
        }
    }

    private fun updateStatus() {
        val uptime = ManagementFactory.getRuntimeMXBean().uptime
        val uptimeStr = formatUptime(uptime)

        val statusTemplates = Jackiro.config.currentlyPlaying
        if (statusTemplates.isEmpty()) return

        val statusTemplate = statusTemplates[currentIndex]
        val shardManager = Jackiro.instance.shardManager

        val filledStatus = statusTemplate
            .replace("{guilds}", shardManager.guilds.size.toString())
            .replace("{users}", shardManager.users.size.toString())
            .replace("{uptime}", uptimeStr)

        val now = System.currentTimeMillis()
        if (now - lastUpdate >= UPDATE_INTERVAL_MS) {
            shardManager.setPresence(OnlineStatus.ONLINE, Activity.playing(filledStatus))
            currentIndex = (currentIndex + 1) % statusTemplates.size
            lastUpdate = now
        }
    }

    private fun formatUptime(uptimeMillis: Long): String {
        val days = TimeUnit.MILLISECONDS.toDays(uptimeMillis)
        val hours = TimeUnit.MILLISECONDS.toHours(uptimeMillis) % 24
        val minutes = TimeUnit.MILLISECONDS.toMinutes(uptimeMillis) % 60
        val seconds = TimeUnit.MILLISECONDS.toSeconds(uptimeMillis) % 60
        return "${days}d ${hours}h ${minutes}m ${seconds}s"
    }
}
